src/material/dialog/dialog-ref.ts

```diff
@@ -122,10 +125,22 @@ export class MatDialogRef<T, R = any> {
       this._closeFallbackTimeout = setTimeout(() => {
         this._overlayRef.dispose();
       }, event.totalTime + 100);
+
+      // Now that the observable close fallback exists clear the backup one
+      clearTimeout(this._destroyCloseFallbackTimeout);
     });
 
     this._containerInstance._startExitAnimation();
     this._state = MatDialogState.CLOSING;
+
+    // Similar to _closeFallbackTimeout but acts as a fallback till the animationStateChanged
+    // observable executes for edge Cases like path changes with dialog open which might not
+    // have time to allow allow any animation events.
+    this._destroyCloseFallbackTimeout = setTimeout(() => {
+      if (this._overlayRef.hasAttached()) {
+        this._overlayRef.dispose();
```

Comment:
I'm not sure how I feel about this, because now we've got the cleanup logic in two separate places. Since we know how long the animation is supposed to take, couldn't we change the observable in `close` to be something like this?

```ts
merge(
  _containerInstance._animationStateChanged.pipe(filter(event => event.phaseName === 'done' && event.toState === 'exit')),
  of(null).pipe(delay(whateverTheAnimationShouldBe + 100))
).subscribe(...);
```

Reply:
@crisbeto Thanks for sharing that. The race function is new to me but cool that it exists.

Just to make sure I understand your approach you are suggesting race() on 2 observables. 1 is the done/exit observable which will contain the total time for that observable to emit. The second is an on-the-fly created observable that emits after 200ms that contains maps to whateverTheAnimationDurationShouldBe which is our guess/static-variable of how long the animation should take.

Did I get those correct? Learning still so I appreciate the feedback if I misunderstood something.

Also are you suggesting this change in the close function to replace the https://github.com/angular/components/blob/09f07c18b8d3956d078547c8a5b418ec4c1ebcfc/src/material/dialog/dialog-ref.ts#L110 observable or the constructor which sets the close animation observable: https://github.com/angular/components/blob/09f07c18b8d3956d078547c8a5b418ec4c1ebcfc/src/material/dialog/dialog-ref.ts#L75

Reply:
Our current setup is waiting for the animation to start in order for the rest of the logic to kick in, but that breaks if the animation never starts. The proposal from above fixes it by adding another observable that'll emit if the animation doesn't start within a certain amount of time. If it does start on time, then it should cancel the fallback observable.

Reply:
Got it. I'll test with that solution.

Reply:
@crisbeto I tested the following solution which appears to work and removes the need for both timeout functions:

`/**
   * Close the dialog.
   * @param dialogResult Optional result to return to the dialog opener.
   */
  close(dialogResult?: R): void {
    this._result = dialogResult;

    // Transition the backdrop in parallel to the dialog.
    this._containerInstance._animationStateChanged.pipe(
      filter(event => event.phaseName === 'start'),
      take(1)
    )
    .subscribe(() => {
      this._beforeClosed.next(dialogResult);
      this._beforeClosed.complete();
      this._state = MatDialogState.CLOSED;
      this._overlayRef.detachBackdrop();
    });

    this._containerInstance._startExitAnimation();
    this._state = MatDialogState.CLOSING;

    // The logic that disposes of the overlay depends on the exit animation completing, however
    // it isn't guaranteed if the parent view is destroyed before that event. Add a fallback
    // observer which will clean everything up if the animation hasn't completed within a specified
    // amount of time plus 100ms. We don't need to run this outside the NgZone, because for the
    // vast majority of cases the timeout will have been cleared before it has the chance to fire.
    race(
      this._containerInstance._animationStateChanged
        .pipe(
          filter(event => event.phaseName === 'done' && event.toState === 'exit'),
          map(event => event.totalTime)
        ),
      of(null).pipe(delay(200), mapTo(200)) //200ms = estimated animation time of 100ms + 100ms
    ).subscribe(() => {
      // Checks if the overlay still exists after assumed animation time has elapsed
      if (this._overlayRef.hasAttached()) {
        // Dispose of overlay assuming animation has not completed in time
        this._overlayRef.dispose();
      }
    });
  }`

If that looks correct to you I can commit the code change later this evening. The only deviation from what you described is that the race function always fires the attached subscribe function so it acts as a fallback for both instances where the animation completes as well as instances where it doesn't. If it does or doesn't is verified by the `if (this._overlayRef.hasAttached())` check. Unless I am overlooking something this seems to work perfectly for all instances.

I'm also not a huge fan of hard coding the fallback animation time of 200ms but I'm not sure how else to get the animation time from the observable itself since there is the case where it may not start yet.

Reply:
@crisbeto The referenced solution is committed to the branch if you wanted to take a look. If needed I can rebase my branch to just the final changes to clean out all the "tweak" commits :)